### DISCUSSION TITLE
Discussions API

### DISCUSSION THREAD

**Discussion by avelino** [community/community#43] [Discussions] [answered]
are there plans to implement api for discussions like [teams discussions](https://developer.github.com/v3/teams/discussions/)?

my need:
I'm moving public communication with my community ([pREST](https://github.com/prest/prest)) to discussions and i want to read the communications to list on the site

**Comment by becca** [comment]
Not yet but we're working towards providing one. This came up as well in https://github.com/github/feedback/discussions/36.

**Reply by smorimoto** [reply]
Will something like `has_discussions` be added here after it's made generally available?
https://docs.github.com/en/free-pro-team@latest/rest/reference/repos#create-an-organization-repository--parameters

**Reply by GMallikarjunaRao** [reply]
I wan to work with Open Mp, MPI and GPU in colab can any one help me

**Reply by draidin135730** [reply]
hi my friends i am Dr.aidin jabariasl . i am veterinary medicine doctor.and nowdays work as second job i want learn and colaburate in github code and programing. i like cryptocurrency issue

**Comment by EliteMasterEric** [comment]
The Discussions appear to be in beta, so it seems like repository discussions will inevitably get an API once the structure of the feature is stable.

**Reply by draidin135730** [reply]
Please send me roadmap

**Comment by vanbroup** [comment]
While any API would already be really helpful, it would be great if you can upload RFC 822 messages similar as how you can import archives to [Google Groups](https://developers.google.com/admin-sdk/groups-migration/v1/guides/manage-email-migrations).

Specifically, we would like to migrate some [Mailman](https://www.gnu.org/software/mailman/) .mbox archives to GitHub discussions.

**Comment by sull** [comment]
Is there an estimated timeline for an API? I am in a decision mode and my preference is to use repo discussion for my need but my workflow would require ability to post a new discussion via API (just the initial post, not full discussion). Alternatively, an API to do the conversion of an Issue to a Discussion would suffice and be a good stopgap solution until full API is ironed out (since I can make an issue via API already).

**Reply by evi-liu** [reply]
Hi @sull! We're hoping to have a private beta available for our Discussions GraphQL API within the next few weeks. Would you like to be part of the beta once it's available?

**Reply by sull** [reply]
great to hear, yes I would like to beta test it. Thanks for letting me know.

**Reply by bpmct** [reply]
I would love to be a part of the beta as well, if possible @evi-liu.

**Reply by matthewmueller** [reply]
I would also love to be part of the beta.

**Reply by msameen99** [reply]
Hi @evi-liu !, I would love to test this as well, thanks in advance

**Reply by josiauh** [reply]
@evi-liu i'd like to test!

**Reply by evi-liu** [reply]
The GraphQL API has been released and anyone may try it out now! See [this commen](https://github.com/github/feedback/discussions/43#discussioncomment-399047)t for details.

**Comment by michaelsena** [comment]
This would be great. We're considering to use Discussions for our open source communities, but it's important that we can use webhooks to grab all discussion events (new, comment, reply, etc) and post them in a channel in our Discord community chat.

**Comment by evi-liu** [comment]
## Private beta for Discussions GraphQL API available

🎉 We now have a private beta available for our Discussions GraphQL API. Please see our [documentation](https://gist.github.com/smashwilson/311e1487ddb40a455fc54d294cc63ad4) for what capabilities are available and note that when making GraphQL requests, you'll need to include the `GraphQL-Features: discussions_api` HTTP header.

Please note that webhooks are not available yet, but our team is actively working on them to make available soon. 

If you'd like to try out the private beta, please reply to this thread and I'll flag you in. I am limiting to the first 15 responders for now.

**Reply by jasonnutter** [reply]
Interested!

**Reply by vanbroup** [reply]
Thanks, is it correct that this API is not intended for importing archive data?

I don't see an option to backdate information?

**Reply by ghost** [reply]
I'm also interested.

**Reply by evi-liu** [reply]
@jasonnutter @TheDancingCode flagged you both in!

@vanbroup where do you intend to import archived data from?

**Reply by psifertex** [reply]
Interested if it gets me the ability to get slack notifications (whether webhook or codeql, I just want ability to trigger outside notifications!)

Edit: after reading the docs, it looks like I'm better off waiting for webhooks. Will wait (mostly patiently) for that. :-)

**Reply by rexledesma** [reply]
@evi-liu hi! Would love to be part of the private beta

**Reply by evi-liu** [reply]
@rexledesma you've now been flagged in!

**Reply by rexledesma** [reply]
hi @evi-liu,

I'm still getting an error regarding this. Here is how I am accessing the beta, after [following the instructions](https://gist.github.com/smashwilson/311e1487ddb40a455fc54d294cc63ad4#feature-flag):

```
import os
from gql import gql, Client
from gql.transport.requests import RequestsHTTPTransport

GITHUB_GRAPHQL_URL = "https://api.github.com/graphql"

token = os.getenv("GITHUB_TOKEN")
client = Client(
    transport=RequestsHTTPTransport(
        url=GITHUB_GRAPHQL_URL,
        headers={"Authorization": f"bearer {token}", "GraphQL-Features": "discussions_api"},
        retries=3
    )
)

query=gql(
    r"""
query {
  repository(owner: "dagster-io", name: "dagster") {
    discussions(first: 10) {
      totalCount
      nodes {
        id
      }
    }
  }
}
    """
)

client.execute(query)
```

However, I'm getting an error 
```
~/.pyenv/versions/3.7.4/envs/internal-3.7.4/lib/python3.7/site-packages/gql/client.py in execute(self, document, *args, **kwargs)
     76         result = self._get_result(document, *args, **kwargs)
     77         if result.errors:
---> 78             raise Exception(str(result.errors[0]))
     79
     80         return result.data

Exception: {'path': ['query', 'repository', 'discussions'], 'extensions': {'code': 'undefinedField', 'typeName': 'Repository', 'fieldName': 'discussions'}, 'locations': [{'line': 3, 'column': 5}], 'message': "Field 'discussions' doesn't exist on type 'Repository'"}
```

**Reply by smashwilson** [reply]
@rexledesma It looks like you hadn't quite made it into the flag somehow. I've re-added you - give it another try now and let us know if you have better luck or not!

**Reply by vanbroup** [reply]
> @vanbroup where do you intend to import archived data from?

@evi-liu would like to import an `mbox` archive from mailman, an [mbox](http://qmail.org./man/man5/mbox.html) file can contain many `RFC 822` formatted messages. While it would be great if we could import an `RFC 822` formatted message directly, [like as in the Google Groups migration API](https://developers.google.com/admin-sdk/groups-migration/v1/guides/manage-email-migrations), I understand this would require more development on the GitHub side. For now, it would help if we could just backdate discussions, comments, and attachments so that the import logic and message binding can bind be handled on the client side.

**Reply by typeofweb** [reply]
@evi-liu I'd like to give it a go as well :) Thanks!

**Reply by evi-liu** [reply]
@mmiszy flagged in!

**Reply by typeofweb** [reply]
@evi-liu could you also add it to my app? I'm testing a new concept here. https://github.com/apps/type-of-web

**Reply by evi-liu** [reply]
@mmiszy just flagged in the app :)

**Reply by typeofweb** [reply]
Thanks!

**Reply by bpmct** [reply]
I'm interested, could you flag me in?

**Reply by molnard** [reply]
I am interested as well, thx 🙏

**Reply by zlliang** [reply]
Interested! Thanks 🙏

**Reply by evi-liu** [reply]
✅ everyone in this thread has been flagged in. please remember to see our [documentation](https://gist.github.com/smashwilson/311e1487ddb40a455fc54d294cc63ad4) for the available capabilities and note that when making GraphQL requests, you'll need to include the `GraphQL-Features: discussions_api HTTP` header.

**Question for you all:** what's everyone using the API for? or what do you intend to use it for?

**Reply by mdidon** [reply]
@evi-liu @smashwilson If this beta remains open, please could I be flagged in?

**Reply by zlliang** [reply]
@evi-liu Seems that I haven't been flagged in, right? Accoring to the [documentation](https://gist.github.com/smashwilson/311e1487ddb40a455fc54d294cc63ad4), I'm trying to make the following request

```js
import { GraphQLClient, gql } from 'graphql-request'

let client = new GraphQLClient('https://api.github.com/graphql', {
  headers: {
    authorization: `bearer ${process.env.GITHUB_TOKEN}`,
    'GraphQL-Features': 'discussions_api'
  }
})

let res = await client.request(gql`
  query {
    repository(owner: "zlliang", name: "discussions-api-playground") {
      discussions {
        totalCount
      }
    }
  }
`)

```

and get

```
ClientError: Field 'discussions' doesn't exist on type 'Repository'
```

Could you please have a check? Thanks a lot! 🙏🙏

**Reply by chrisblakley** [reply]
If/when a webhooks beta of the Discussions API becomes available, I would be interested in that. I do not need a seat for the GraphQL beta (to save space for others). Thanks.

**Reply by johnlindquist** [reply]
Please flag me in :D

**Reply by smashwilson** [reply]
@zlliang and @johnlindquist: just flagged you both in 🏁

**Reply by zlliang** [reply]
> @zlliang and @johnlindquist: just flagged you both in 🏁

Thanks!

**Reply by draidin135730** [reply]
i am agree with you

**Reply by nul800sebastiaan** [reply]
@evi-liu I'd like in as well! 
Usage would be to expose discussions and updates to discussions to our community site, so people visiting there get a change to see what's going on and can chime in.
Additionally, we have a custom dashboard to show our team what's new and unanswered, we can see what needs attention in a central hub for all of our repositories.

**Reply by laymonage** [reply]
@nul800sebastiaan The API is already available to all users: https://docs.github.com/en/graphql/guides/using-the-graphql-api-for-discussions

**Reply by nul800sebastiaan** [reply]
@laymonage Ha! That's perfect, thanks! I didn't read far enough down 🤦‍♂️ Sorry about that and thanks for the heads up! 🎉

**Reply by mktcode** [reply]
@nul800sebastiaan @laymonage  But I believe you still need to set the header: `"GraphQL-Features": "discussions_api"` (not sure though, can't test it right now)

**Reply by laymonage** [reply]
@mktcode No, that's no longer necessary. You even left a comment on the [announcement](https://github.com/github/feedback/discussions/3622#discussioncomment-821148) 😉

I'm using it on [my project](https://giscus.app) and it works fine.

**Reply by mktcode** [reply]
lol :D

**Reply by josiauh** [reply]
okok,

**Reply by dperolio** [reply]
Any new info available on the webhooks situation?

**Comment by sull** [comment]
Please flag me

**Reply by evi-liu** [reply]
flagged you in!

**Comment by whydnxx** [comment]
Please flag me 😉

**Reply by evi-liu** [reply]
@whydnxx flagged in!

**Comment by evi-liu** [comment, marked as answer]
## 🎉 Discussions GraphQL API now available!

Great news, everyone! The Discussions GraphQL API is now shipped and available for use. See our [documentation](https://docs.github.com/en/graphql/guides/using-the-graphql-api-for-discussions) for how to get started.

**Reply by PranamBhat** [reply]
**Wow!!!**

**Reply by fredysomy** [reply]
I have a doubt..
As said in the docs we need a CategoryId to do the graphql query right..
Is there a way to get it..?

**Reply by smashwilson** [reply]
You can find category IDs by querying the `discussionCategories` connection from a `Repository`, like this:

```graphql
{
  repository(owner: "smashwilson", name: "discussions-test") {
    discussionCategories(first: 10) {
      nodes {
        id
        name
      }
    }
  }
}

```

**Reply by mktcode** [reply]
I'd also love to test this new API. Our project could really benefit from it. I also wonder how to access the votes. Are they handled as reactions?

**Reply by laymonage** [reply]
The public schema does not yet include the Discussions API. If anyone needs the schema, you can get it using:

```bash
curl \
    -H "Authorization: bearer $GITHUB_TOKEN" \
    -H "Accept: application/vnd.github.v4.idl" \
    -H "GraphQL-Features: discussions_api" \
    https://api.github.com/graphql | \
jq -r .data > github.graphql
```

Don't forget to set the `$GITHUB_TOKEN` variable.

**Reply by mktcode** [reply]
Yes, I figured that out by now. ;) We're experimenting! Discussions for Organizations would also be nice.

**Reply by NiklasEi** [reply]
Am I the only one running into CORS issues when I set the `GraphQL-Features` header? See https://github.community/t/cors-issue-with-graphql-features-header/167899

Is it correct that there is no way of getting a public discussion without a Token at the moment?

Will discussions also be added to the REST API?

**Reply by laymonage** [reply]
@mktcode I don't know if you meant organization-wide Discussions or not, but [Team Discussions](https://docs.github.com/en/github/setting-up-and-managing-organizations-and-teams/about-team-discussions) is a thing (even before repo Discussions).

@NiklasEi I haven't run into such an issue, but I haven't tried creating the request from the browser.

**Reply by laymonage** [reply]
@NiklasEi I can confirm, I encountered the same issue when trying to use the API from https://hoppscotch.io.

**Reply by mktcode** [reply]
@laymonage Exactly, I mean organization-wide Discussions.... that are also accessible for the community. Maybe I missed something in the Team Discussions but I couldn't find how to make them public.

![image](https://user-images.githubusercontent.com/6792578/111762407-44c7f780-88a1-11eb-968c-306630ef941b.png)
Seems not to be intended to be public.

I'd like to have something here:
![image](https://user-images.githubusercontent.com/6792578/111762544-688b3d80-88a1-11eb-8cc9-690f8a1dde51.png)

I think a third option in this upper screenshot would be enough, an then based whether there are public discussions or not, display a Discussions tab on the Org's page. Well... actually not. I just realized that they don't at all share the same functionality. The team discussions feature is much simpler.

**Reply by laymonage** [reply]
@mktcode Well, in that case, I guess you could create an empty repository specifically for Discussions. Not ideal, but at least you can have a centralized discussion.

**Reply by donggyu04** [reply]
@smashwilson 

Hi, I try your example with curl but, It emits `undefinedField` errors. Is there any wrong with my request?
thanks in advance

#### 1. Request
`curl -H "Authorization: token MY_GITHUB_TOKEN" -H "Content-Type: application/json" -X POST -d '{ "query": "query { repository(owner: \"OWNER\", name: \"MY_REPONAME\") { discussionCategories(first: 10) { nodes { id name } } } }" }' https://api.github.com/graphql`


#### 2. Response

`{"errors":[{"path":["query","repository","discussionCategories"],"extensions":{"code":"undefinedField","typeName":"Repository","fieldName":"discussionCategories"},"locations":[{"line":1,"column":70}],"message":"Field 'discussionCategories' doesn't exist on type 'Repository'"}]}`

![image](https://user-images.githubusercontent.com/14273601/113965220-ecfd2a80-9867-11eb-9ea0-b6f4d100b6d3.png)

**Reply by laymonage** [reply]
@donggyu04 As noted in the documentation,

> To use this API, you must include `GraphQL-Features: discussions_api` in the `HTTP` header.

**Reply by donggyu04** [reply]
@laymonage 

Oh, It is completely my mistake 😭 thanks!
But, If possible, It would be better if the error message was more clear about the reason like "Required header 'GraphQL-Features' is missing" :)

**Reply by Bibo-Joshi** [reply]
Hi. Great that Discussions are available via the GraphQL api. However, are there any plans to make the available via the rest API as well?

**Reply by rockrzhackrz** [reply]
How to enable GraphQL-Features: discussions_api in http ....?

Iam using graphql explorer https://docs.github.com/en/graphql/overview/explorer


Iam getting error when i use the above queries 


`{
  "errors": [
    {
      "path": [
        "query",
        "repository",
        "discussionCategories"
      ],
      "extensions": {
        "code": "undefinedField",
        "typeName": "Repository",
        "fieldName": "discussionCategories"
      },
      "locations": [
        {
          "line": 5,
          "column": 5
        }
      ],
      "message": "Field 'discussionCategories' doesn't exist on type 'Repository'"
    }
  ]
}`
> 
> 
> @donggyu04 As noted in the documentation,
> 
> > To use this API, you must include `GraphQL-Features: discussions_api` in the `HTTP` header.

**Reply by rockrzhackrz** [reply]
@donggyu04 

iam getting this error even though iam using correct

![errorz](https://user-images.githubusercontent.com/58619807/117475427-a63f4500-af79-11eb-92cb-e40196613d42.png)

**Reply by rockrzhackrz** [reply]
@donggyu04 what is the tool used at 1:18 in this video https://www.youtube.com/watch?v=pftVGtUswZM THanks!

**Reply by SuspiciousLookingOwl** [reply]
I might've missed something from the [doc](https://docs.github.com/en/graphql/guides/using-the-graphql-api-for-discussions#discussion), but how do I get the up vote count of a `Discussion` / `DiscussionComment`?

**Reply by mktcode** [reply]
Same question here. I actually just assumed it would be handled as reactions but I just checked that and I was wrong. Can't find this information anywhere.

**Reply by mktcode** [reply]
@rockrzhackrz I found this: https://lucasconstantino.github.io/graphiql-online/

**Reply by laymonage** [reply]
@SuspiciousLookingOwl @mktcode The docs aren't up to date as they're still actively developing the API. For example, you used to be able to upvote and downvote comments, and you'd use something like a `netUpvoteCount`. Now they only allow you to upvote, and you can access the value using the `upvoteCount` attribute in a `Discussion`/`DiscussionComment`.

For the latest version of the API, you can download the schema using the method I've shared [here](https://github.com/github/feedback/discussions/43#discussioncomment-494848).

Here's the latest one for `DiscussionComment`, you can see there's an `upvoteCount: Int!` field in it.

<details>
<summary><code>type DiscussionComment</code></summary>

```graphql
"""
A comment on a discussion.
"""
type DiscussionComment implements Comment & Deletable & Minimizable & Node & Reactable & Updatable & UpdatableComment & Votable {
  """
  The actor who authored the comment.
  """
  author: Actor

  """
  Author's association with the subject of the comment.
  """
  authorAssociation: CommentAuthorAssociation!

  """
  The body as Markdown.
  """
  body: String!

  """
  The body rendered to HTML.
  """
  bodyHTML: HTML!

  """
  The body rendered to text.
  """
  bodyText: String!

  """
  Identifies the date and time when the object was created.
  """
  createdAt: DateTime!

  """
  Check if this comment was created via an email reply.
  """
  createdViaEmail: Boolean!

  """
  Identifies the primary key from the database.
  """
  databaseId: Int

  """
  The time when this replied-to comment was deleted
  """
  deletedAt: DateTime

  """
  The discussion this comment was created in
  """
  discussion: Discussion

  """
  The actor who edited the comment.
  """
  editor: Actor
  id: ID!

  """
  Check if this comment was edited and includes an edit with the creation data
  """
  includesCreatedEdit: Boolean!

  """
  Has this comment been chosen as the answer of its discussion?
  """
  isAnswer: Boolean!

  """
  Returns whether or not a comment has been minimized.
  """
  isMinimized: Boolean!

  """
  The moment the editor made the last edit
  """
  lastEditedAt: DateTime

  """
  Returns why the comment was minimized.
  """
  minimizedReason: String

  """
  Identifies when the comment was published at.
  """
  publishedAt: DateTime

  """
  A list of reactions grouped by content left on the subject.
  """
  reactionGroups: [ReactionGroup!]

  """
  A list of Reactions left on the Issue.
  """
  reactions(
    """
    Returns the elements in the list that come after the specified cursor.
    """
    after: String

    """
    Returns the elements in the list that come before the specified cursor.
    """
    before: String

    """
    Allows filtering Reactions by emoji.
    """
    content: ReactionContent

    """
    Returns the first _n_ elements from the list.
    """
    first: Int

    """
    Returns the last _n_ elements from the list.
    """
    last: Int

    """
    Allows specifying the order in which reactions are returned.
    """
    orderBy: ReactionOrder
  ): ReactionConnection!

  """
  The threaded replies to this comment.
  """
  replies(
    """
    Returns the elements in the list that come after the specified cursor.
    """
    after: String

    """
    Returns the elements in the list that come before the specified cursor.
    """
    before: String

    """
    Returns the first _n_ elements from the list.
    """
    first: Int

    """
    Returns the last _n_ elements from the list.
    """
    last: Int
  ): DiscussionCommentConnection!

  """
  The discussion comment this comment is a reply to
  """
  replyTo: DiscussionComment

  """
  The path for this discussion comment.
  """
  resourcePath: URI!

  """
  Identifies the date and time when the object was last updated.
  """
  updatedAt: DateTime!

  """
  Number of upvotes that this subject has received.
  """
  upvoteCount: Int!

  """
  The URL for this discussion comment.
  """
  url: URI!

  """
  A list of edits to this content.
  """
  userContentEdits(
    """
    Returns the elements in the list that come after the specified cursor.
    """
    after: String

    """
    Returns the elements in the list that come before the specified cursor.
    """
    before: String

    """
    Returns the first _n_ elements from the list.
    """
    first: Int

    """
    Returns the last _n_ elements from the list.
    """
    last: Int
  ): UserContentEditConnection

  """
  Check if the current viewer can delete this object.
  """
  viewerCanDelete: Boolean!

  """
  Can the current user mark this comment as an answer?
  """
  viewerCanMarkAsAnswer: Boolean!

  """
  Check if the current viewer can minimize this object.
  """
  viewerCanMinimize: Boolean!

  """
  Can user react to this subject
  """
  viewerCanReact: Boolean!

  """
  Can the current user unmark this comment as an answer?
  """
  viewerCanUnmarkAsAnswer: Boolean!

  """
  Check if the current viewer can update this object.
  """
  viewerCanUpdate: Boolean!

  """
  Whether or not the current user can add or remove an upvote on this subject.
  """
  viewerCanUpvote: Boolean!

  """
  Reasons why the current viewer can not update this comment.
  """
  viewerCannotUpdateReasons: [CommentCannotUpdateReason!]!

  """
  Did the viewer author this comment.
  """
  viewerDidAuthor: Boolean!

  """
  Whether or not the current user has already upvoted this subject.
  """
  viewerHasUpvoted: Boolean!
}
```
</details>

**Reply by SuspiciousLookingOwl** [reply]
@laymonage `upvoteCount` works, terima kasih!

**Reply by surjithctly** [reply]
Is there any plans for the REST API? @evi-liu ?

**Reply by Lite5h4dow** [reply]
rest api?

**Comment by laymonage** [comment]
It is currently impossible to use the API from a web browser. See https://github.com/github/feedback/discussions/3622.

**Reply by surjithctly** [reply]
Waiting for REST API 🤞

**Reply by pattacini** [reply]
Any news regarding the REST API?

**Reply by Bibo-Joshi** [reply]
https://github.com/github/feedback/discussions/4327#discussioncomment-1616032

**Reply by pattacini** [reply]
Ouch.. that's a pity.
Thanks for posting!

**Reply by aleks-ivanov** [reply]
Assuming you could do the same GET/POST actions with GraphQL, I see that as a win 🙂

**Reply by pattacini** [reply]
Well, when you have an extensive codebase written in REST API (not limited to this request) that you need to port to GrapQL sooner or later is not a win in the short term 😏

**Comment by starsalon** [comment]
> are there plans to implement api for discussions like [teams discussions](https://developer.github.com/v3/teams/discussions/)?
> 
> my need:
> I'm moving public communication with my community ([pREST](https://github.com/prest/prest)) to discussions and i want to read the communications to list on the site

I am trying to implement api for my website https://epesa.in/. It is payment platform. I need help.

**Reply by philip-gai** [reply]
Here is an example of me using the graphql api from a node app with typescript:
1. [Get repo discussion categories](https://github.com/philip-gai/announcement-drafter/blob/main/web-app/src/services/githubService.ts#L84-L107)
2. [Create a repo discussion](https://github.com/philip-gai/announcement-drafter/blob/main/web-app/src/services/githubService.ts#L192-L221)
3. [Create an org team discussion](https://github.com/philip-gai/announcement-drafter/blob/main/web-app/src/services/githubService.ts#L59-L73)

This was for my GitHub App [Announcement Drafter](https://github.com/marketplace/repo-st) that allows you to draft and create discussions using pull requests!

**Comment by aleks-ivanov** [comment]
I would LOVE a Discussions API, similar to the Issues one 🙂

**Reply by aleks-ivanov** [reply]
> [#4327 (comment)](https://github.com/github/feedback/discussions/4327#discussioncomment-1616032)

**Comment by devstein** [comment]
Hi all 👋  I read through this thread and have two questions

1. Is there a way to convert an issue to discussion or vice-versa via the GraphQL API?
2. Is there a way to know if a discussion was created from an issue via the GraphQL or REST API?

Any help is appreciated!

**Reply by abhi-iac** [reply]
I am in a same boat. Banged my head searching through gh REST documentation for discussion REST API. Now that I can see it is not available I might search for the GraphQL options.
